File: domain/src/main/java/net/janczar/powertape/example/domain/usecase/GetMessagesUseCase.java
package net.janczar.powertape.example.domain.usecase;


import net.janczar.powertape.annotation.Inject;
import net.janczar.powertape.annotation.Provide;
import net.janczar.powertape.example.domain.UseCase;
import net.janczar.powertape.example.domain.UseCaseComposer;
import net.janczar.powertape.example.domain.model.Message;
import net.janczar.powertape.example.domain.repository.MessagesRepository;

import java.util.List;

import io.reactivex.Observable;

public class GetMessagesUseCase extends UseCase<Void, List<Message>> {

    @Inject
    MessagesRepository messagesRepository;

    @Provide
    public GetMessagesUseCase(final UseCaseComposer composer) {
        super(composer);
    }

    @Override
    protected Observable<List<Message>> createUseCaseObservable(Void param) {
        return messagesRepository.getMessages();
    }
}

File: domain/src/main/java/net/janczar/powertape/example/domain/usecase/GetMessagesUseCase.kt
package net.janczar.powertape.example.domain.usecase

import io.reactivex.Single
import net.janczar.powertape.annotation.Provide
import net.janczar.powertape.example.domain.SchedulersProvider
import net.janczar.powertape.example.domain.VoidSingleUseCase
import net.janczar.powertape.example.domain.model.Message
import net.janczar.powertape.example.domain.repository.MessagesRepository


class GetMessagesUseCase @Provide constructor(
        schedulersProvider: SchedulersProvider,
        private val messagesRepository: MessagesRepository
): VoidSingleUseCase<List<Message>>(schedulersProvider) {
    override fun createUseCaseSingle(): Single<List<Message>> {
        return messagesRepository.getMessages()
    }
}